lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php

```diff
@@ -123,41 +130,76 @@ public function convertException($message, DriverException $exception)
 
     /**
      * {@inheritdoc}
+     *
+     * @return AbstractPlatform|MariaDb102Platform|MySQL57Platform|MySqlPlatform
+     * @throws DBALException
      */
     public function createDatabasePlatformForVersion($version)
     {
-        if ( ! preg_match('/^(?P<major>\d+)(?:\.(?P<minor>\d+)(?:\.(?P<patch>\d+))?)?/', $version, $versionParts)) {
-            throw DBALException::invalidPlatformVersionSpecified(
-                $version,
-                '<major_version>.<minor_version>.<patch_version>'
-            );
+        if (false !== stripos($version, 'mariadb')
+            && version_compare($this->getMariaDbMysqlVersionNumber($version), '10.2.7', '>=')) {
+            return new MariaDb102Platform();
         }
 
-        if (false !== stripos($version, 'mariadb')) {
-            return $this->getDatabasePlatform();
+        if (false === stripos($version, 'mariadb')
+            && version_compare($this->getOracleMysqlVersionNumber($version), '5.7.9', '>=')) {
+            return new MySQL57Platform();
         }
 
+        return $this->getDatabasePlatform();
+    }
+
+    /**
+     * Get a normalized 'version number' from the server string
+     * returned by Oracle MySQL servers.
+     *
+     * @param string $versionString Version string returned by the driver, i.e. '5.7.10'
+     * @throws DBALException
+     */
+    private function getOracleMysqlVersionNumber(string $versionString) : string
+    {
+        if (!preg_match(self::MYSQL_ORACLE_VERSION_REGEXP, $versionString, $versionParts)) {
+            throw DBALException::invalidPlatformVersionSpecified(
+                $versionString,
+                '<major_version>.<minor_version>.<patch_version>'
+            );
+        }
         $majorVersion = $versionParts['major'];
-        $minorVersion = isset($versionParts['minor']) ? $versionParts['minor'] : 0;
-        $patchVersion = isset($versionParts['patch']) ? $versionParts['patch'] : null;
+        $minorVersion = $versionParts['minor'] ?? 0;
+        $patchVersion = $versionParts['patch'] ?? null;
 
         if ('5' === $majorVersion && '7' === $minorVersion && null === $patchVersion) {
             $patchVersion = '9';
         }
 
-        $version = $majorVersion . '.' . $minorVersion . '.' . $patchVersion;
+        return $majorVersion . '.' . $minorVersion . '.' . $patchVersion;
+    }
 
-        if (version_compare($version, '5.7.9', '>=')) {
-            return new MySQL57Platform();
+    /**
+     * Detect MariaDB server version, including hack for some mariadb distributions
+     * that starts with the prefix '5.5.5-'
+     *
+     * @param string $versionString Version string as returned by mariadb server, i.e. '5.5.5-Mariadb-10.0.8-xenial'
+     * @throws DBALException
+     */
+    private function getMariaDbMysqlVersionNumber(string $versionString) : string
+    {
+        $version = str_replace('5.5.5-', '', $versionString);
```

Comment:
I'd rather have this as part of the regexp directly, this looks hacky.

Reply:
Previously discussed before.

It looks hacky because it is ;) More importantly, It's good to let it like this till we figure out if we can unify the regexp for maria and mysql.

> @morozov: According to the comment, the client just needs to strip the '5.5.5-' prefix, so I don't see the need in a special parser. Maybe we can just unify the logic of parsing MySQL and MariaDB versions so that we always parse the string representation, not the integer (because it doesn't work with all supported platforms).

Something like:

```
$serverInfo = $this->_conn->get_server_info();

if (false !== stripos($serverInfo, 'mariadb')) {
    $serverInfo = str_replace('5.5.5-', '', $serverInfo);
}
```

Reply:
This is what I was suggesting:
`'/^(?:5\.5\.5-)?(mariadb-)?(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)/'`

Reply:
Mixed feelings about it, but I'm going to trust you for that :) Changed accordingly.

Reply:
@Majkl578 also note that I've reflected the change in the exception message as well:

```php
        if (!preg_match('/^(?:5\.5\.5-)?(mariadb-)?(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)/i', $versionString, $versionParts)) {
            throw DBALException::invalidPlatformVersionSpecified(
                $versionString,
                '^(?:5\.5\.5-)?(mariadb-)?<major_version>.<minor_version>.<patch_version>'
            );
        }
```

Just wondering if it's not too much :D Let me know and I'll revert